primitive/src/main/java/io/atomix/primitive/impl/AbstractAsyncPrimitiveProxy.java

```diff
@@ -0,0 +1,239 @@
+   * @return A completable future to be completed with the operation result. The future is guaranteed to be completed after all
+   * {@link PrimitiveOperation} submission futures that preceded it.
+   * @throws NullPointerException if {@code operation} is null
+   */
+  protected CompletableFuture<Void> acceptOn(PartitionId partitionId, Consumer<S> operation) {
+    return getServiceProxy(partitionId).accept(operation);
+  }
+
+  /**
+   * Submits an empty operation to the given partition.
+   *
+   * @param partitionId the partition in which to execute the operation
+   * @param operation   the operation identifier
+   * @param <R>         the operation result type
+   * @return A completable future to be completed with the operation result. The future is guaranteed to be completed after all
+   * {@link PrimitiveOperation} submission futures that preceded it.
+   * @throws NullPointerException if {@code operation} is null
+   */
+  protected <R> CompletableFuture<R> applyOn(PartitionId partitionId, Function<S, R> operation) {
+    return getServiceProxy(partitionId).apply(operation);
+  }
+
+  /**
+   * Submits an empty operation to the owning partition for the given key.
+   *
+   * @param key       the key for which to submit the operation
+   * @param operation the operation
+   * @return A completable future to be completed with the operation result. The future is guaranteed to be completed after all
+   * {@link PrimitiveOperation} submission futures that preceded it.
+   * @throws NullPointerException if {@code operation} is null
+   */
+  protected CompletableFuture<Void> acceptBy(String key, Consumer<S> operation) {
+    return getServiceProxy(key).accept(operation);
+  }
+
+  /**
+   * Submits an empty operation to the owning partition for the given key.
+   *
+   * @param key       the key for which to submit the operation
+   * @param operation the operation
+   * @param <R>       the operation result type
+   * @return A completable future to be completed with the operation result. The future is guaranteed to be completed after all
+   * {@link PrimitiveOperation} submission futures that preceded it.
+   * @throws NullPointerException if {@code operation} is null
+   */
+  protected <R> CompletableFuture<R> applyBy(String key, Function<S, R> operation) {
+    return getServiceProxy(key).apply(operation);
+  }
+
+  /**
+   * Service proxy container.
+   */
+  private class ServiceProxy<S> {
+    private final S proxy;
+    private final ServiceProxyHandler handler;
+
+    ServiceProxy(S proxy, ServiceProxyHandler handler) {
+      this.proxy = proxy;
+      this.handler = handler;
+    }
+
+    /**
+     * Invokes a void method on the underlying proxy.
+     *
+     * @param operation the operation to perform on the proxy
+     * @return the resulting void future
+     */
+    CompletableFuture<Void> accept(Consumer<S> operation) {
+      operation.accept(proxy);
+      return handler.getResultFuture();
```

Comment:
@johnou this is an example of the proxy being used and then the future being immediately read from it.